corehq/apps/user_importer/importer.py

```diff
@@ -582,35 +578,63 @@ def create_or_update_commcare_users_and_groups(upload_domain, user_specs, upload
+                        was_group_change = True
                         group.remove_user(user)
 
                 for group_name in group_names:
-                    domain_info.group_memoizer.by_name(group_name).add_user(user, save=False)
+                    if group_name not in user_curr_group_names:
+                        user_added_to_groups.append(domain_info.group_memoizer.by_name(group_name))
+                        was_group_change = True
+                        domain_info.group_memoizer.by_name(group_name).add_user(user, save=False)
+
+                try:
+                    for domain_info in domain_info_by_domain.values():
+                        domain_info.group_memoizer.save_all()
+                except BulkSaveError as e:
+                    _error_message = (
+                        "Oops! We were not able to save some of your group changes. "
+                        "Please make sure no one else is editing your groups "
+                        "and try again."
+                    )
+                    logging.exception((
+                        'BulkSaveError saving groups. '
+                        'User saw error message "%s". Errors: %s'
+                    ) % (_error_message, e.errors))
+                    ret['errors'].append(_error_message)
+
+                if was_group_change:
+                    updated_groups = []
+                    for group in user_curr_groups:
+                        if group not in removed_from_groups:
+                            updated_groups.append(group)
+                    updated_groups.extend(user_added_to_groups)
+                    commcare_user_importer.logger.add_info(
+                        UserChangeMessage.groups_info(updated_groups)
+                    )
+                    print("updated groups:", updated_groups)
+                    print("get groups:", domain_info.group_memoizer.by_user_id(user.user_id))
+
+                user.save()
+                commcare_user_importer.save_log()
+
             except ValidationError as e:
                 status_row['flag'] = e.message
             except (UserUploadError, CouchUser.Inconsistent) as e:
                 status_row['flag'] = str(e)
 
             ret["rows"].append(status_row)
     finally:
-        try:
-            for domain_info in domain_info_by_domain.values():
-                domain_info.group_memoizer.save_all()
-        except BulkSaveError as e:
-            _error_message = (
-                "Oops! We were not able to save some of your group changes. "
-                "Please make sure no one else is editing your groups "
-                "and try again."
-            )
-            logging.exception((
-                'BulkSaveError saving groups. '
-                'User saw error message "%s". Errors: %s'
-            ) % (_error_message, e.errors))
-            ret['errors'].append(_error_message)
-
-    return ret
+        return ret
```

Comment:
is this correct indentation?

Reply:
I think it's correct. It caught my eye too when I first moved it into that `finally` block, but all local and automated testing checks out.

Reply:
since there is no code in the `finally` block now you could remove the entire `try / finally`. Having the return statement in there will not make the function return the data if an exception is thrown.

Reply:
I was wrong about the `return` statement FYI. If you `return` from inside a finally block when an exception has been raised then the exception will get silenced which is even worse.